### ISSUE TITLE
🤝  Implement AWS Transfer Family Server 

### ISSUE THREAD

**Issue by Gary-H9** [ministryofjustice/analytical-platform#3501]
### User Story

As a… User of the Analytical Platform
I want to be able to ingest data into the platform
So that… I can use all of the tooling etc within the platform

As a AP Product Engineer 
I want to the platform to be able to ingest data in a controlled, precise and monitored fashion.
So that we can provide a better service for our users and build a foundational offering within the platform.

This ticket builds on this previously raised [Feature Request](https://github.com/ministryofjustice/data-platform/issues/799). 

### Value / Purpose

Data ingestion will be a foundational part of the AP offering going forward. This piece of work will create the foundations which this offering will be built on. 

### Useful Contacts

Jacob W / Julia / Gary

### Proposal

Create the ingestion route as outlined [here](https://github.com/ministryofjustice/modernisation-platform/issues/6364).

### Additional Information

![image](https://github.com/ministryofjustice/data-platform/assets/26419401/db697b0b-a852-4ed4-b42c-b8d585a74fa4)

For the AWS Transfer Family Server - [Electronic Monitoring ](https://github.com/ministryofjustice/modernisation-platform-environments/tree/main/terraform/environments/electronic-monitoring-data)already use this functionality.

### Definition of Done

- [x] 📝 Documentation has been written / updated
- [x] Modernisation Platform AWS Account created ([WIP](https://github.com/ministryofjustice/modernisation-platform/issues/6364))
- [x] AWS Transfer Family Server Terraform'ed
- [x] IAM Roles defined + Terraform'ed
- [x] User testing arranged 
- [x] Testing completed

**Comment by jacobwoffenden** [comment]
29/02/24 summary:

* @Gary-H9 requested new Modernisation Platform environment https://github.com/ministryofjustice/modernisation-platform/issues/6364

**Comment by julialawrence** [comment]
@YvanMOJdigital

**Comment by jacobwoffenden** [comment]
05/03/24 summary:

- @jacobwoffenden #3602
- @Gary-H9 
  - initial configuration of ingestion scanning container - [WIP PR](https://github.com/ministryofjustice/analytical-platform-ingestion-scanner/pull/1)
  - Addition of CloudWatch EventBridge, lambda function and associated testings of the above in concert with these
  - Modernisation Platform environment build in progress

**Comment by jacobwoffenden** [comment]
06/03/24:

- Successfully tested flow for updating and then consuming ClamAV definitions
- Successfully tested scanning files using S3 bucket notifications
  - Clean files moved to a processed bucket, while infected are moved to quarantine
    - Files are deleted from landing once moved


TODO:
- [x] Add object policy to deny access to quarantined
- [x] Configure Transfer Family server
  - [x] Home directory
  - [x] Move into VPC **and** give an EIP
  - [x] Custom hostname?
  - [x] Restrict users to specific IP ranges
- [ ] ~Notification for quarantined objects~
  - ~Can we use a tag from the transfer user to send an email? and in the future a message to AP dashboard~ Split into #3770 

Notes:
- Electronic Monitoring create a server **per** provider, we don't want to do this, could be hard to scale and manager via AP dashboard 

EDIT @jacobwoffenden:

I've managed to get this working by editing the IAM policy for the user to include more S3 permissions, and added KMS permissions

<img width="1076" alt="Screenshot 2024-03-06 at 18 37 42" src="https://github.com/ministryofjustice/data-platform/assets/15148673/289ecf2f-b8c3-4db0-9dbc-61df561c3ee0">

EDIT 2 @jacobwoffenden:

- [ ] ~figure out how/where to run `aws transfer update-server` to set `DirectoryListingOptimization` to `ENABLED`. this is not exposed in Terraform~ MP ClickOps'd it

**Comment by Gary-H9** [comment]
07/03/24: 
- Converted the lambda function to bash from python. Noted faster runs as a result. Tested both scenarios successfully. 
- Planned architecture relating to notifications in variety of scenarios.

**Comment by jacobwoffenden** [comment]
11/04/24 summary:

- More work on scan functionality
  - sends a message to SNS when quarantining a file
- Started work on notify functionality
  - Triggers on SNS publish
  - Plumbed into Data Platform's GOV.UK Notify account, have tested basic functionality
- Drafted architecture for "supplier data", or information about supplier we don't want to store in public Terraform, this will be superseded by AP dashboard eventually
- Liaised with GDS about getting access to Analytical Platform's notify account

TODO:
- [ ] ~SNS redrive logic~
- [x] clarify naming/terminology (data contact, data owner, etc.)

**Comment by Gary-H9** [comment]
14th March summary: 
- Refactored existing code + added code for new lambda and required components
- Removed calls to SNS for this iteration of the product - this can be looked at after MVP - [ticket created for this](https://github.com/ministryofjustice/data-platform/issues/3770)
- Created ingestion-transfer `handler.py` 
- Created a test bucket in `analytical-platform-dev` called `dev-ingestion-testing` to test cross account file move from the above lambda. Manually updated the policy on this but it needs further work.

**Comment by jacobwoffenden** [comment]
18/03/2024 summary (plus a bit more):

- https://github.com/ministryofjustice/modernisation-platform-environments/pull/5371
- https://github.com/ministryofjustice/modernisation-platform-environments/pull/5400 

Gary: 
- continued development on the ingestion-transfer image.

**Comment by Gary-H9** [comment]
- Closed development [branch/PR](https://github.com/ministryofjustice/data-platform/pull/3553) in `data-platform` repository
- Migrating [code](https://github.com/ministryofjustice/modernisation-platform-environments/pull/5419) from the above branch into Modernisation Platform Environments repository. [Merged](https://github.com/ministryofjustice/modernisation-platform-environments/actions/runs/8362847004) - development and production built.
- Updated DNS record for both environments in `analytical-platform-production` R53.

**Comment by Gary-H9** [comment]
Created documentation relating to the solution in [user-guidance](https://github.com/moj-analytical-services/user-guidance/pull/402) (🚧) and in our new [runbooks documentation](https://runbooks.analytical-platform.service.justice.gov.uk/documentation/runbooks/001-ingestion-maintenance.html).

**Comment by jacobwoffenden** [comment]
@ministryofjustice/modernisation-platform have enabled optimised directories manually in both transfer servers (dev and prod)

**Comment by jacobwoffenden** [comment]
https://github.com/hashicorp/terraform-provider-aws/issues/35851

**Comment by Gary-H9** [comment]
Awaiting user information to allow testing. In the meantime [egress](https://arc.net/l/quote/hthdpttu) has been completed.

**Comment by Gary-H9** [comment]
- [Locked down S3 Bucket Permissions on the Quarantine bucket](https://github.com/ministryofjustice/modernisation-platform-environments/pull/5533)
- [Egress functionality](https://github.com/ministryofjustice/modernisation-platform-environments/pull/5503/commits/262937f41e3e05db3497a82de854b25c7f9dc879#diff-c665e819a1025538f3201fa33f8253d80ff0421ca30f10ac1851b5046bf23e85)
- No update from users RE testing.

**Comment by jacobwoffenden** [comment]
Pending details from BOLD to begin onboarding

**Comment by jacobwoffenden** [comment]
Data Engineering's https://github.com/moj-analytical-services/iam_builder needs updating to add `kms` which is needed to add KMS permissions to their Airflow role

**Comment by jacobwoffenden** [comment]
Analytical Platform team to update Airflow IAM role with permissions to access KMS key

- [x] `arn:aws:iam::593291632749:role/airflow_dev_bold_rr_essex_police`

**Comment by jacobwoffenden** [comment]
Implemented regional KMS

- https://github.com/ministryofjustice/data-platform/pull/4106
- https://github.com/ministryofjustice/data-platform/pull/4108

**Comment by jacobwoffenden** [comment]
Moving to blocked:

- ~BOLD Airflow role needs access to KMS, release made to IAM Build (https://github.com/moj-analytical-services/iam_builder/releases/tag/v4.3.0)~
- DE's `mojap-land` S3 buckets need their policy updating to allow ingestion service

**Comment by michaeljcollinsuk** [comment]
Moving back to in progress as @julialawrence is working on a new request from BOLD

**Comment by jacobwoffenden** [comment]
Blocked by #3765

**Comment by jacobwoffenden** [comment]
Closing as we've tested end-to-end, we're waiting on BOLD to perform their end-to-end testing which is out of scope for this issue.
